### ISSUE TITLE
Update Previous Model with New Categories

### ISSUE THREAD

**Issue by ronils428** [huggingface/setfit#372]
Is there a way to add categories based on new data?

For example - Initially I trained a model with 5 categories and saved the model. I now have new data that I want to feed into the model but this new data has 8 categories. Would I have to start from scratch or can I use the original model I trained?

Thank you!

**Comment by tomaarsen** [comment]
Hello!

You'd need to update the shape of the classifier. The classifier is usually relatively simple quick to train (as compared to the sentence-transformer finetuning), meaning that it may be possible to reuse the sentence-transformer (`model.model_body`) but retrain the classifier from scratch (`model.model_head`). 

Consider the following example:

```python
from datasets import load_dataset
from sklearn.linear_model import LogisticRegression
from setfit import SetFitModel, SetFitTrainer, sample_dataset

### Training a model with 2 classes - standard stuff so far
# Load a dataset from the Hugging Face Hub
old_dataset = load_dataset("sst2")

# Simulate the few-shot regime by sampling 8 examples per class
old_train_dataset = sample_dataset(old_dataset["train"], label_column="label", num_samples=16)
old_eval_dataset = old_dataset["validation"]

# Load a SetFit model from Hub
old_model = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2")

# Create trainer
old_trainer = SetFitTrainer(
    model=old_model,
    train_dataset=old_train_dataset,
    eval_dataset=old_eval_dataset,
    column_mapping={"sentence": "text", "label": "label"},
)

# Train and evaluate
old_trainer.train()
metrics = old_trainer.evaluate()
print("sst2 validation performance:", metrics)
# => 88.65% accuracy

### This is where you saved your old model (5 categories), and from here we'll continue.
### In your example, you would only have the code below here, but with this line too:
# old_model = SetFitModel.from_pretrained("path/to/old/model")

### Modifying the 2 class model to 5 classes
# Load a dataset from the Hugging Face Hub
dataset = load_dataset("SetFit/sst5")

train_dataset = sample_dataset(dataset["train"], label_column="label", num_samples=16)
x_train = train_dataset["text"]
y_train = train_dataset["label"]
eval_dataset = dataset["validation"]

# Take the old trained sentence-transformer
sentence_transformer = old_model.model_body
# Make a new classifier and fit it
classifier = LogisticRegression()
x_embeddings = sentence_transformer.encode(x_train)
classifier.fit(x_embeddings, y_train)
# Create a model using the initializer
model = SetFitModel(sentence_transformer, classifier)
# Make a Trainer - only for evaluation
trainer = SetFitTrainer(
    model=model,
    eval_dataset=eval_dataset,
)
metrics = trainer.evaluate()
print("sst5 validation performance:", metrics)
# => 39.42% accuracy
```
Note that here we only fit the classifier - this should be really quick.

It also seems to outperform fitting using a non-finetuned sentence-transformer:
```python
from datasets import load_dataset
from sklearn.linear_model import LogisticRegression
from setfit import SetFitModel, SetFitTrainer, sample_dataset

# Load a dataset from the Hugging Face Hub
dataset = load_dataset("SetFit/sst5")

train_dataset = sample_dataset(dataset["train"], label_column="label", num_samples=16)
x_train = train_dataset["text"]
y_train = train_dataset["label"]
eval_dataset = dataset["validation"]

model = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2")
x_embeddings = model.model_body.encode(x_train)
model.model_head.fit(x_embeddings, y_train)
trainer = SetFitTrainer(
    model=model,
    eval_dataset=eval_dataset,
)
metrics = trainer.evaluate()
print("sst5 validation performance:", metrics)
# => 37.60%
```

However, it's much worse than fully retraining (in my case):
```python
from datasets import load_dataset
from setfit import SetFitModel, SetFitTrainer, sample_dataset

dataset = load_dataset("SetFit/sst5")

train_dataset = sample_dataset(dataset["train"], label_column="label", num_samples=16)
eval_dataset = dataset["validation"]

model = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2")
trainer = SetFitTrainer(
    model=model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)
trainer.train()
metrics = trainer.evaluate()
print("sst5 validation performance:", metrics)
# => 43.50%
```

Roughly equivalently, you can continue training using your old model instead:
```python
from datasets import load_dataset
from sklearn.linear_model import LogisticRegression
from setfit import SetFitModel, SetFitTrainer, sample_dataset

### Training a model with 2 classes - standard stuff so far
# Load a dataset from the Hugging Face Hub
old_dataset = load_dataset("sst2")

# Simulate the few-shot regime by sampling 8 examples per class
old_train_dataset = sample_dataset(old_dataset["train"], label_column="label", num_samples=16)
old_eval_dataset = old_dataset["validation"]

# Load a SetFit model from Hub
old_model = SetFitModel.from_pretrained("sentence-transformers/paraphrase-mpnet-base-v2")

# Create trainer
old_trainer = SetFitTrainer(
    model=old_model,
    train_dataset=old_train_dataset,
    eval_dataset=old_eval_dataset,
    column_mapping={"sentence": "text", "label": "label"},
)

# Train and evaluate
old_trainer.train()
metrics = old_trainer.evaluate()
print("sst2 validation performance:", metrics)
# => 88.65% accuracy

### This is where you saved your old model (5 categories), and from here we'll continue.
### In your example, you would only have the code below here, but with this line too:
# old_model = SetFitModel.from_pretrained("path/to/old/model")

### Modifying the 2 class model to 5 classes
# Load a dataset from the Hugging Face Hub
dataset = load_dataset("SetFit/sst5")

train_dataset = sample_dataset(dataset["train"], label_column="label", num_samples=16)
eval_dataset = dataset["validation"]

trainer = SetFitTrainer(
    model=old_model,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)
trainer.train()
metrics = trainer.evaluate()
print("sst5 validation performance:", metrics)
# => 43.32% accuracy
```

In short, I would recommend retraining the entire model, either using your old model as the initial checkpoint, or using a "fresh" sentence-transformer. However, with this comment, you should be able to figure out how to retrain just the classifier and reuse the sentence-transformer for a quick-to-train model with a bit less performance.

- Tom Aarsen

**Comment by tomaarsen** [comment]
I think this has been answered! I'll close this for now.

- Tom Aarsen
